core/src/main/scala/cromwell/core/retry/Retry.scala

```diff
@@ -0,0 +1,53 @@
+package cromwell.core.retry
+
+import akka.actor.ActorSystem
+import cromwell.core.CromwellFatalException
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.concurrent.{ExecutionContext, Future}
+import akka.pattern.after
+
+object Retry {
+  /**
+    * Retries a Future on a designated backoff strategy until either a designated number of retries or a fatal error
+    * is reached.
+    *
+    * @param f A Future thunk which will be executed once per cycle
+    * @param maxRetries An optional number of times to retry the thunk. If this is None, there is no limit.
+    * @param backoff An exponential backoff strategy to use for each retry strategy.
+    * @param isTransient An optional function of Throwable => Boolean. If provided and it returns true, this throwable
+    *                    does not count towards the maxRetries limit.
+    * @param isFatal An optional function of Throwable => Boolean. If provided and it returns true, withRetry will halt
+    * @param actorSystem An ActorSystem, used for its scheduler and dispatcher.
+    * @tparam A The return type of the thunk
+    * @return The final completed Future
+    */
+  def withRetry[A](f: => Future[A],
+                   maxRetries: Option[Int] = Option(10),
+                   backoff: SimpleExponentialBackoff = SimpleExponentialBackoff(5 seconds, 10 seconds, 1.1D),
+                   isTransient: Option[Throwable => Boolean] = None,
+                   isFatal: Option[Throwable => Boolean] = None)
```

Comment:
ToL: The `Option` wrapper plus the implicit utility below seem like overkill. Can you cite precedent in any project / codebase for a parameter that is an option-of-a-function? I'd even accept an overloaded java method with similar semantics. ;) I know that null is a "billion dollar mistake", but is it so bad to just have these parameters return `false` by default instead of `None`?